### DISCUSSION TITLE
Rewrite and seperate Pyloxone

### DISCUSSION THREAD

**Discussion by JoDehli** [JoDehli/PyLoxone#290] [Announcements]
## 👋 Welcome!

I want to separate Pyloxone in 2 packages. One should be responsible for the communication and the other should be the custom integration for home assistant. Ideally the communication package can be used for other python based packages which need to interact with the loxone miniserver. 

Pro: 
- the communication module can be used for other projects
- the home assistant custom integration is much smaller and maybe it can become a core module for home assistant

Cons:
- development becomes harder as you have to maintain 2 projects
- dependency management can be difficult 

For this I already created a seperate Github Project. https://github.com/JoDehli/pyloxone-api
But this implementation is not really reliable. I also started a new attempt which is nearly finished but not public available for now.

**Comment by JoDehli** [comment]
@gigatexel If you want I can make the pyloxone Branch public and make a new branch on https://github.com/JoDehli/pyloxone-api where I can commit the latest version of the 2 branches.

**Comment by gigatexel** [comment]
Good Idea!

Op vr 2 aug 2024 om 15:47 schreef Joachim D. ***@***.***>

**Comment by JoDehli** [comment]
@gigatexel I started a new branch (which is still private) where I started the API implementation. But it still in a state where I can not share it. Sorry I do not really have time at the moment. A lot of do at work and other projects at home. Hope I can make the branch public at the end of the year.

**Comment by Jakob-Gliwa** [comment]
@JoDehli I used your Loxone WS API as base for my implemention for [loxinflux](https://github.com/Jakob-Gliwa/loxInFlux )and a not yet publicly published mqtt-relay (basically a docker alternative to the Loxberry mqtt-gateway, which I'm still testing).

Changed I did:
- Generalized the access through a more generic callback-mechanism
- Consolidated "websockets" and "httpx" libraries into aiohttp (which in my benchmarking was faster for websockets)
- tested with uvloop
- Did some things to increase performance (esp. ported to use orjson for json parsing, wrote custom Cython parser for type2 and type3 messages as these flood the ws the most when updates are activated)
- Generally thinned out the api class and split with other classes
- Also tested the parsing of all messages types with "construct", but havent fully published that as it is not applicable for my use cases

But there are also changes to main design patterns of your original, like mostly relying on exceptions rather then Error-Codes (as this seems to be more in line with aiohttps design logic).

There'd be some todos to package it as separate module:
- Finish parsing of all other complex message types (e.g. Daytime)
- Documentation
- Maybe rethinking the returned message format as it is currently not coherent between message types
- Generally having more pairs of eyes on the implemenation...

I also thought about publishing it separatelly but didn't see it as appropriate as it is heavily built upon your work and still has much of your flavor in it.

Feel free to take a look into it, maybe it is of interest for you!

**Reply by JoDehli** [reply]
@Jakob-Gliwa good the someone uses my code for other projects. As I already wrote I rewrote all the websocket stuff completely new but had no time to really release and test it. The actual implementation is very bad in my opinion but time is the biggest problem. I do not know how long I can keep this repo up to date as my time is really limited at the moment. So I concentrate only on the most important things not on any new features. That is why I can not implement your improvements at the moment but I will take a look when I have more time.

**Comment by JoDehli** [comment]
@gigatexel 
@Jakob-Gliwa 
As I have emphasized multiple times, I am not satisfied with the current implementation of the integration. Therefore, I have been thinking about how the work on the integration could continue in the future.

I don’t like that the connection to the Miniserver is integrated into Home Assistant. That’s why I have decided to attempt implementing everything via MQTT. Home Assistant will then receive all information from MQTT, similar to the Zigbee2MQTT project.

Anyone interested can take a look at the initial version. So far, it only supports publishing Loxone commands to MQTT.

https://github.com/JoDehli/Pyloxone2Mqtt

**Reply by gigatexel** [reply]
But that would require some other system running the MQTT implementation, effectively requiring differt hardware since HAOS does not support running things in parallel?
Or maybe we could make Pyloxone2Mqtt run as a HA-Addon?

**Reply by JoDehli** [reply]
@gigatexel of course this should run as add-on in Homeassistant like zigbee2mqtt. Do you know that? It should also have an admin website where you can decide which entity should added to Homeassistent

**Reply by JoDehli** [reply]
@gigatexel most important thing should be to make the reconnection stable so that Homeassistent does get errors if loxobe restarts or the connection is unstable. I had many complaints about the connection in the past. And also all the Homeassistent stuff which I had to update because they changed something. 

And a big advantage can be that you can template a mqtt sensor or device first manually and if you know it works you can implement an automatic creation for similar devices.

**Reply by Jakob-Gliwa** [reply]
On the topic of reconnection:
- For the MQTT client: Have a look at gmqtt. I found it's built in reconnection to work more reliable than asynciomqtt (whilst also being more efficient).
- For the websocket part: This was really a pain, since the keepalive alone wasn't really enough. In my adaption of your websocket client implementation I added a dedicated ping to check whether we are still online or not, to be super sure to reconnect properly

On the overall topic - i think in an ideal world we'd have a "ecosystem" of separate tools/packages as follows:
1. A robust, performant **Websocket-Client** package/instance, which builds the base for many of the following features
2. A **brige** publishing loxone websocket events to MQTT
3. A **relay** MQTTxLoxoneMiniserver for sending and receiving MQTT commands
4. A robust miniserver **backup service**
5. A connector to **MQTTxHASS** (either as HASS-Addon or as separate module - maybe even integrated in 2. using the MQTT-device pattern of HASS)
6. A **TICK-stack connector** (best case a native telegraf input-addon)
7. A dockerized **Musicserver**

I think we are close to that goal since we now have solutions for 2, 3, 4, 6 and WIPs for 1 and 5 !!!

**Reply by JoDehli** [reply]
@Jakob-Gliwa I do not understand why you'd do the whitelist stuff. I would pass everything to mqtt and let Homeassistent or any other system what only on the messages they need.

**Reply by Jakob-Gliwa** [reply]
The whitelist is just for inbound messages (inbound from the miniserver perspective).

E.g. if you subscribe to zigbee2mqtt/# in theory the relay would try to forward all messages from zigbee2mqtt to the miniserver.
But the miniserver is just interested in commands that have a corresponding VirtualInput. So I implemented the whitelist to filter out all messages that would result in errormessages anyways to not bombard the miniserver with irrelevant requests.

There are two ways you can use the whitelist in the relay:
1. Set it manually
2. Enable the "sync with miniserver" option - here the mqtt relay is pulling the relevant information from the miniserver config and populates the whitelist based on that

Outbound commincation is always send as-is.

But I think here is the difference between the relay and Pyloxone2Mqtt. I think you are forwarding all messages from the websocket to mqtt, right? The relay just sends commands issued on a discrete VirtualOut.It's basically the Loxberry MQTT Gateway if you ever came across that one.

My other project https://github.com/Jakob-Gliwa/loxInFlux is probably closer to what you want to achieve. This one more or less sends every call from the websocket to a channel of your choice (stout, tcp, udp, mqtt). But it right now it does so in influx line protocol style. I think the websocket-client and mqtt-client implementations of that project could be used more or less as-is for that functionality. Also the logic that extracts more of the relevant control informations (categories, rooms, uuid-subcontrols, etc.) could also be helpful to a generic Websocket-to-MQTT Bridge.

**Reply by JoDehli** [reply]
@Jakob-Gliwa so you need to setup a virtual output in loxone? 


I think we are not very different. 
For me the ideal implementation would be to pass all events from websocket to mqtt. Everything that can talk to mqtt can listen to the event. The event bus listens to the mqtt event which would pass it to loxone. This is done only thru websocket. All messages and all Infos are transferred thru websocket. 

On top of these it should be possible to make addons to do the translation between the mqtt messages to the special format for the device which the adson is used for. For example Homeassistant or openhab or.......


But I at the beginning. This is all my head. I spent about 3 hours on the current version. So was not a lot of work. Maybe I test your code and if it works for me too I can cancel the work on my repo.

**Reply by gigatexel** [reply]
> @gigatexel of course this should run as add-on in Homeassistant like zigbee2mqtt. Do you know that? It should also have an admin website where you can decide which entity should added to Homeassistent

Just quicky make a simple addon that fetches the repository on Github. All parameters can be specified in HA, a .env file is generated and the code is launched :-)

![image](https://github.com/user-attachments/assets/86ff6bf4-1ba2-4171-a8b8-b7e989ec063d)
![image](https://github.com/user-attachments/assets/93f609d1-ec12-4757-9872-74b67222bf21)
![image](https://github.com/user-attachments/assets/5d701630-236e-4da1-838a-f109e8424f07)

**Comment by Jakob-Gliwa** [comment]
Haha seems as if great minds think alike :D 

Just a couple of days ago i published my take on the same topic: [https://github.com/Jakob-Gliwa/loxMqttRelay](url)

It's a bidirectional bridge between Loxone and MQTT.
I'm in betatesting with a user and my own installation right now - it runs stable in my environment since about christmas.

Right now I've rewritten the core in Rust (about 300x the performance of python for the whole filtering part etc.). Just struggeling with one or two pyo3-asyncio integration parts, that would make it even more coherent.

**Reply by JoDehli** [reply]
@Jakob-Gliwa I think the performance should be no problem with python. Of course rust is much faster but async programming in rust is a pain. Golang  would be the best for this. 

But it does not matter from what source the mqtt events come.

**Reply by JoDehli** [reply]
@Jakob-Gliwa it is sad that we have not worked together for an implementation. We have so many good projects and so many projects which are not supported anymore.

**Reply by Jakob-Gliwa** [reply]
Yeah really a shame. But I think I misinterpreted what you are trying to achieve. You want to publish Loxone websocket events to MQTT, right? I just want to give Loxone Miniserver the ability to communicate over mqtt (e.g. trigger devices via zigbee2mqt or on HASS). So at least we didn't do the same work twice :D 

Wouldn't it be enought to use the existing node-red implementation and couple it with mqtt? I think the Node implementation might even be more suited for that use case then python would be.

With regards to the performance:
- For the websocket client I noticed, that, if you are in the mode that sends you all events, you really get tons of updates especially from Energy-Meters. With Python alone I was at abou 2-3% of CPU usage on my M2 Macbook and 12600k server - I know, not much. But if it runs in parallel with several other containers it sums up unnecessarily
- For the mqtt relay: There are scenarios were you might receive quite big JSON documents (e.g. if you run a ems-esp). Again one alone wouldn't break anything but the more of such devices there are in your network the more you'll notice it

But yeah I also bite my a** about not doing it in Golang from day one. Would have made the Telegraf integration also easier...

**Reply by JoDehli** [reply]
@Jakob-Gliwa I want not only to send data to mqtt. I want to replace my current implementation of Pyloxone with the new one. Like zigbee2mqtt. With an event bus and async it should absolutely be no problem to send thousands of messages through websocket and mqtt.

So I think it is the same what we want to achieve. my goal would be to make a plugin system to add more translators to other systems.

**Comment by gigatexel** [comment]
Are we aiming a using MQTT Discovery to create the devices/entities in HA?
If so, this would make the work a lot easier to bridge between MQTT and HA. It's just a matter of following the specs in creating messages.

**Comment by JoDehli** [comment]
@gigatexel Yes this was my idea. As I said it is a test. But as you can see the addon works. ;-) 

1. I the first step should be to make the reconnect for both mqtt and loxone rock solid. 
2. I would prefer to make a separate homassistant.py where it creates the mqtt devices by mqtt discovery. Of course sensor should be no problem but the real fun comes with light controllers etc. 
3. We should make a website internal in the docker or addon to configure stuff like friendly name. I would prefer to only add the devices to homassistant that I really need. Not all. 

It should make the work easier because:
 - no dependencies with homeassistant (we can use whatever we want and what is good)
 - new devices can be testet by making a mqtt template entry or even not supported devices
 - connection is not interrupted when loxone restarts or homeassistant restarts

**Reply by gigatexel** [reply]
> @gigatexel Yes this was my idea. As I said it is a test. But as you can see the addon works. ;-)
> 
> 1. I the first step should be to make the reconnect for both mqtt and loxone rock solid.
> 2. I would prefer to make a separate homassistant.py where it creates the mqtt devices by mqtt discovery. Of course sensor should be no problem but the real fun comes with light controllers etc.
> 3. We should make a website internal in the docker or addon to configure stuff like friendly name. I would prefer to only add the devices to homassistant that I really need. Not all.
> 
> It should make the work easier because:
> 
> * no dependencies with homeassistant (we can use whatever we want and what is good)
> * new devices can be testet by making a mqtt template entry or even not supported devices
> * connection is not interrupted when loxone restarts or homeassistant restarts

If someone could create a demo device implementation ( for instance a simple button), I’m willing to start implementing other loxone blocks according to the Mqtt autodiscovery spec.

**Comment by JoDehli** [comment]
@Jakob-Gliwa 
@gigatexel 
Maybe @Jakob-Gliwa can help us with the docker setup and publishing if the software is ready for the first release. 

Of course this should take a while. It is really a first test. ;-)

**Reply by gigatexel** [reply]
> @Jakob-Gliwa @gigatexel Maybe @Jakob-Gliwa can help us with the docker setup and publishing if the software is ready for the first release.
> 
> Of course this should take a while. It is really a first test. ;-)

The way I set up my Addon demo is that it takes the code from the repository on rebuild and it stays compatible with the .env setup. In that way, we can easily test by just rebuilding - no need to pull code manually from GitHub. If you want, I can publish the Dockerfile, config and run.sh

**Reply by JoDehli** [reply]
@gigatexel of course this would be great. We should keep in mind that it should be possible to run it natively, on docker and as Homeassistent addon. The Homeassistent addon can be a extra repo which uses also the original git repo. I think the addon for Homeassistent usage should be on top of the pure  mqtt communication.

**Reply by JoDehli** [reply]
@gigatexel do you want to be a maintainer of the branch.

**Reply by gigatexel** [reply]
> @gigatexel do you want to be a maintainer of the branch.

Sure, fine for me
Maybe we should move this discussion to the other repository?

**Comment by gigatexel** [comment]
Before I put more effort into something that no one approves :-)

I modified _do_start_listening in lib/pyloxone_api/connection.py so the full LoxAPP3.json gets published on the event bus

<img width="562" alt="image" src="https://github.com/user-attachments/assets/003df7b4-6059-43f6-b9d5-c5cfc12f2ee5" />

I created a homeassistant.py file that listens for the creation of the LoxAPP3 topic, and when it gets created, it just parses it into different topics according to the JSON.
My idea was to take this as a starting point for translating all the info in LoxAPP3.json into the MQTT Autodiscovery format for the different devices/entities.


<img width="561" alt="image" src="https://github.com/user-attachments/assets/73143ebb-e9c7-4637-8b9a-5a19e7cc722a" />

**Reply by Jakob-Gliwa** [reply]
If you need a shortcut look at: https://github.com/Jakob-Gliwa/loxInFlux/blob/main/src/loxInFlux/miniserver.py

It consumes the loxone miniserver config and extracts all controls with their metadata, so that its easier to e.g. map name, category, rooms, etc. to the control uuid

Pros:
- More metadata per control incl. der subcontrols
- Having really ALL controls, not just the ones visible via the visu -> Allows for a grabber pattern, to also publish the state and controlling of non-visu-visible controls

Cons:
- Miniserver-User needs to have FTP rights on Miniserver

**Reply by JoDehli** [reply]
@gigatexel inwould Stick with the normal json which the loxone apps get. So no consuming the the ftp config.

**Comment by gigatexel** [comment]
Maybe one of the easiest, but all switches are imported as a device+switch and they respond to actual state changes. No attributes, rooms, etc yet.

https://github.com/user-attachments/assets/93815e04-d2f7-4704-ade8-295d77e2c9ed

**Reply by JoDehli** [reply]
@gigatexel yes this looks good. At the moment I have no time to work on this due to some health problems. But as soon as I can I want to continue with the work. 

What I think is really important that everything goes over the event bus. And the Homeassistent.py should only be to create the devices. It should not have any logic. That is important.

**Reply by JoDehli** [reply]
@gigatexel I think switches and sensors are not very hard to implement. But the light controllers and other complicated devices could be hard to implement.

What I want is a web gui to decide what loxone entry becomes a Homeassistent mqtt device. Like I zigbee2mqtt. So you use zigbee2mqtt

**Reply by gigatexel** [reply]
> @gigatexel yes this looks good. At the moment I have no time to work on this due to some health problems. But as soon as I can I want to continue with the work.
> 
> What I think is really important that everything goes over the event bus. And the Homeassistent.py should only be to create the devices. It should not have any logic. That is important.

I hope you get well soon! When you do, could you enable discussions in the new repo?

**Reply by gigatexel** [reply]
> @gigatexel I think switches and sensors are not very hard to implement. But the light controllers and other complicated devices could be hard to implement.
> 
> What I want is a web gui to decide what loxone entry becomes a Homeassistent mqtt device. Like I zigbee2mqtt. So you use zigbee2mqtt

Well, we have to start somewhere :-)
I understand what the final goal should be, but just to make it somewhat manageable, I'm trying to break things up:

- Post the LoxAPP3.json to the event bus, so the homeassistant.py can start it's work
- homeassistant.py parses the LoxAPP3.json as a Dict and generates all the Autodiscovery stuff

The HA Autodiscovery docs are not that great and it involves a lot of trial and error.
But now I have created a JSON template for a switch that makes sure the switch gets discovered and that it listens to loxone2mqtt/ topics on its own - so after creation homeassistant.py no longer intervenes. Next step is the comms the other way around and scaling up devices

**Reply by JoDehli** [reply]
@gigatexel I enabled the discussions. 

For the Matt discovery maybe we can look into zigbee2mqtt.

**Reply by gigatexel** [reply]
> @gigatexel I enabled the discussions.
> 
> For the Matt discovery maybe we can look into zigbee2mqtt.

Thx! I Will check it out. Get Well soon!

**Reply by JoDehli** [reply]
@gigatexel perfect would be that each device type could be like a addon. So a be device would only add code to its own python file. 

And for testing or creating a new device it should be possible to make a mqtt template device in Homeassistant itself. I think especially for complicated devices like the light controller it would make sense to test the Jinja template before start with the mqtt discovery.

**Reply by JoDehli** [reply]
> > @gigatexel I enabled the discussions.
> > 
> > For the Matt discovery maybe we can look into zigbee2mqtt.
> 
> Thx! I Will check it out. Get Well soon!


@gigatexel thanks. Hope next weekend I can help you.
